hail/src/main/scala/is/hail/expr/ir/ForwardRelationalLets.scala

```diff
@@ -1,81 +1,83 @@
 package is.hail.expr.ir
 
+import is.hail.backend.ExecuteContext
+import is.hail.utils.fatal
+
 import scala.collection.mutable
+import scala.util.control.NonFatal
 
 object ForwardRelationalLets {
-  def apply(ir0: BaseIR): BaseIR = {
-
-    val usages = mutable.HashMap.empty[Name, (Int, Int)]
-
-    val nestingDepth = NestingDepth(ir0)
-
-    def visit(ir1: BaseIR): Unit = {
-      ir1 match {
-        case RelationalLet(name, _, _) =>
-          usages(name) = (0, 0)
-        case RelationalLetTable(name, _, _) =>
-          usages(name) = (0, 0)
-        case RelationalLetMatrixTable(name, _, _) =>
-          usages(name) = (0, 0)
-        case RelationalLetBlockMatrix(name, _, _) =>
-          usages(name) = (0, 0)
+  def apply(ctx: ExecuteContext, ir0: BaseIR): BaseIR =
+    ctx.time {
+      val uses = mutable.HashMap.empty[Name, (Int, Int)]
+      val nestingDepth = NestingDepth(ctx, ir0)
+      IRTraversal.levelOrder(ir0).foreach {
```

Comment:
One change in this file is to only insert into `uses` when there's a `ReleationalRef`.
The test for `shouldForward` is adjusted for the case when `uses` is empty